src/lib.rs

```diff
@@ -61,10 +69,81 @@ impl BinaryQF {
     pub fn discriminant(&self) -> BigInt {
         // for negative delta we compute 4ac - b^2
         let abs_delta = BigInt::from(4) * &self.a * &self.c - &self.b * &self.b;
-        assert!(abs_delta > BigInt::zero());
+//        assert!(abs_delta > BigInt::zero());
         -abs_delta
     }
 
+    pub fn compressing_form(&self)->BinaryQF_CompressedForm {
+        let t = BinaryQF::compressing_b(&self.a,&self.b);
+        BinaryQF_CompressedForm{a:self.a.clone(), b_compressed:t.clone(), delta:  self.discriminant()}
+    }
+
+    pub fn decompressing_form(com_form:BinaryQF_CompressedForm) ->Self{
+        let a = com_form.a.clone();
+        let comp_b = com_form.b_compressed.clone();
+        let delta:BigInt = com_form.delta.clone();
+        let r_square:BigInt = comp_b.clone()*comp_b.clone()*delta.clone() % a.clone();
+
+        let new_r = r_square.sqrt() %a.clone();
+        let inv_t = BinaryQF::inv(&comp_b.clone(),&a);
+        let b =(new_r.clone()*inv_t) % a.clone();
+        let c = (&b.pow(2) - &delta) / (BigInt::from(4) * &a);
+        BinaryQF { a, b, c }
+    }
+
+
+    pub fn inv(a: &BigInt, n:&BigInt) -> BigInt
```

Comment:
why not use BigInt::inv(...) here?

Reply:
I missed that function (I am not familiar with Mpz)...